### ISSUE TITLE
Error loading market1501 osnet_x1_0 weights under Same-domain ReID in model zoo

### ISSUE THREAD

**Issue by mikel-brostrom** [KaiyangZhou/deep-person-reid#482]
Hi!

I am trying to load the pretrained market1501 `osnet_x1_0` weights under https://kaiyangzhou.github.io/deep-person-reid/MODEL_ZOO.html#same-domain-reid by:

```python

datamanager = torchreid.data.ImageDataManager(
    root='reid-data',
    sources='market1501',
    targets='market1501',
    height=256,
    width=128,
    batch_size_train=32,
    batch_size_test=100,
    transforms=['random_flip', 'random_crop']
)
model = models.build_model(name='osnet_x1_0', num_classes=datamanager.num_train_pids)
torchreid.utils.load_pretrained_weights(model, weight_path='/x/osnet_x1_0_market_256x128_amsgrad_ep150_stp60_lr0.0015_b64_fb10_softmax_labelsmooth_flip.pth')
```

but I get:

```bash
** The following layers are discarded due to unmatched keys or layer size: ['classifier.weight', 'classifier.bias']
```

The weights name has market in it so I supposed it was trained on 1501 classes but this is not the case. Anybody knows that the problem is here?

**Comment by mikel-brostrom** [comment]
I inspected the last layer in the pth file by:

```python
checkpoint = torch.load('/x/osnet_x1_0_market_256x128_amsgrad_ep150_stp60_lr0.0015_b64_fb10_softmax_labelsmooth_flip.pth')
last_key = list(checkpoint)[-1]
print(checkpoint[last_key].size())
```

and what I get is 751. Anybody knows what is going on?

**Comment by sholevs66** [comment]
Hi. I'm just starting to work with this package also.

Can you please tell me where did you find 
`osnet_x1_0_market_256x128_amsgrad_ep150_stp60_lr0.0015_b64_fb10_softmax_labelsmooth_flip` ?
Is this supposed to be a osnet model pretrained on ImageNet followd by training on market1501 train?

I have only managed to load osnet_x1_0_imagenet.pth

Thanks

**Comment by mikel-brostrom** [comment]
> Can you please tell me where did you find `osnet_x1_0_market_256x128_amsgrad_ep150_stp60_lr0.0015_b64_fb10_softmax_labelsmooth_flip` ?

Click on `94.2 (82.6)` under https://kaiyangzhou.github.io/deep-person-reid/MODEL_ZOO.html#same-domain-reid

> Is this supposed to be a osnet model pretrained on ImageNet followd by training on market1501 train?

In the OSNet [paper](https://arxiv.org/pdf/1910.06827.pdf) they state: ` For fair comparison against existing methods, we implement two versions of OSNet. One is trained from  scratch  while  the  other  is  fine-tuned  from  ImageNet pre-trained weights.`. Neither the mAP score nor the R1 score under https://kaiyangzhou.github.io/deep-person-reid/MODEL_ZOO.html#same-domain-reid corresponds to any of the two model results in the paper, so it is hard to tell.
`

I just assumed that this is because this .pth doesn't contain the last classifying re-id layer. am I correct?`

The key for the last layer is `classifier.bias` as in

** The following layers are discarded due to unmatched keys or layer size: ['classifier.weight', 'classifier.bias']

so it must have a classifying ReID layer, but its size is 751 for some reason

**Comment by sholevs66** [comment]
> > Can you please tell me where did you find `osnet_x1_0_market_256x128_amsgrad_ep150_stp60_lr0.0015_b64_fb10_softmax_labelsmooth_flip` ?
> 
> Click on `94.2 (82.6)` under https://kaiyangzhou.github.io/deep-person-reid/MODEL_ZOO.html#same-domain-reid
> 
> > Is this supposed to be a osnet model pretrained on ImageNet followd by training on market1501 train?
> 
> In the OSNet [paper](https://arxiv.org/pdf/1910.06827.pdf) they state: ` For fair comparison against existing methods, we implement two versions of OSNet. One is trained from scratch while the other is fine-tuned from ImageNet pre-trained weights.`. Neither the mAP score nor the R1 score under https://kaiyangzhou.github.io/deep-person-reid/MODEL_ZOO.html#same-domain-reid corresponds to any of the two model results in the paper, so it is hard to tell.
> 
> > I just assumed that this is because this .pth doesn't contain the last classifying re-id layer. am I correct?
> 
> The key for the last layer is `classifier.bias` as in
> 
> `** The following layers are discarded due to unmatched keys or layer size: ['classifier.weight', 'classifier.bias']`
> 
> so it must have a classifying ReID layer, but its size is 751 for some reason

Well, there are 751 identities for train and 751 (or 750, bit confused also) for test. S it fits. Eitherway, if you are using a trained model for eval on the test set, you are ignoring the final classification layer, instead you take the 512-d embedding vector for matching with the gallery. hope it helps is..

**Comment by mikel-brostrom** [comment]
> you are ignoring the final classification layer, instead you take the 512-d embedding vector for matching with the gallery

You are totally right there @sholevs66. 

Then I guess it would be better to use the FeatureExtractor
right away:

```python
extractor = FeatureExtractor(
    model_name='osnet_x1_0',
    model_path='a/b/c/model.pth.tar',
    device='cuda'
)

image_list = [
    'a/b/c/image001.jpg',
    'a/b/c/image002.jpg',
    'a/b/c/image003.jpg',
    'a/b/c/image004.jpg',
    'a/b/c/image005.jpg'
]

features = extractor(image_list)
print(features.shape) # output (5, 512)
```

instead of 

``` python
model = models.build_model(name='osnet_x1_0', num_classes=datamanager.num_train_pids)
```
